lib/IDE/Refactoring.cpp

```diff
@@ -4757,6 +4755,163 @@ struct CallbackClassifier {
+      Collector.walk(Scope);
+    else
+      Collector.walk(SF);
+  }
+
+  /// Collect all explicit references in \p Target that are not first declared.
+  static void collect(ASTNode Target, llvm::DenseSet<const Decl *> &Decls) {
+    ReferenceDeclCollector Collector(Decls);
+    if (Target)
+      Collector.walk(Target);
+  }
+
+private:
+  ReferenceDeclCollector(ASTNode Target, SourceManager *SM,
+                         llvm::DenseSet<const Decl *> &Decls)
+      : SM(SM), DeclaredDecls(), ReferencedDecls(Decls), Target(Target),
+        AfterTarget(false) {}
+
+  ReferenceDeclCollector(llvm::DenseSet<const Decl *> &Decls)
+      : SM(nullptr), DeclaredDecls(), ReferencedDecls(Decls), Target(nullptr),
+        AfterTarget(true) {}
+
+  bool walkToDeclPre(Decl *D, CharSourceRange Range) override {
+    if (AfterTarget && !D->isImplicit()) {
+      DeclaredDecls.insert(D);
+    } else if (Target && D == Target.dyn_cast<Decl *>()) {
+      AfterTarget = true;
+    }
+    return AfterTarget || containsTarget(D->getSourceRange());
+  }
+
+  bool walkToExprPre(Expr *E) override {
+    if (AfterTarget && !E->isImplicit()) {
+      if (auto *DRE = dyn_cast<DeclRefExpr>(E)) {
+        if (auto *D = DRE->getDecl()) {
+          if (!D->isImplicit() && !DeclaredDecls.count(D))
+            ReferencedDecls.insert(D);
+        }
+      }
+    } else if (E == Target.dyn_cast<Expr *>()) {
+      AfterTarget = true;
+    }
+    return AfterTarget || containsTarget(E->getSourceRange());
+  }
+
+  bool walkToStmtPre(Stmt *S) override {
+    if (S == Target.dyn_cast<Stmt *>())
+      AfterTarget = true;
+    return AfterTarget || containsTarget(S->getSourceRange());
+  }
+
+  bool containsTarget(SourceRange Range) {
+    return SM && Target &&
+        SM->rangeContainsTokenLoc(Range, Target.getStartLoc());
+  }
+};
+
+/// This is intended to collect all the declarations for which the async
+/// refactoring should make sure not to redeclare or shadow when hoisting
+/// the body of the completion handler out of its call. It's an over
+/// approximation, but that's better than causing a redeclaration error or
+/// shadowing an existing variable.
+static llvm::DenseSet<const Decl *> collectUsedDecls(ASTNode Target,
+                                                     SourceFile &SF) {
+  llvm::DenseSet<const Decl *> Decls;
+
+  if (auto *D = Target.dyn_cast<Decl *>()) {
+    if (auto *AFD = dyn_cast<AbstractFunctionDecl>(D)) {
+      // Converting a function, check for any references to unknown declartions
+      // (ie. those not defined in the function itself) and also add the
+      // function's parameters. Including the parameters is an over approximation,
+      // but is an easy way to ensure that references to parameters after hoisted
+      // calls are not shadowed by hoisted declarations.
+      for (auto *PD : AFD->getParameters()->getArray()) {
+        Decls.insert(PD);
+      }
+      ReferenceDeclCollector::collect(Target, Decls);
```

Comment:
Is this asking why the reference collector is needed? If so...

The main issue with the conversion is the fact that we hoist the closure body out into the body of the function, possibly even removing some ifs/switches. This means that we can either cause redeclarations or (IMO worse) shadowing.

Redeclarations are easy enough to find and the reference collector isn't needed for those. It's purely for the shadowing case, ie. see the `shadowsUsedDecl` test case (added some comments here):
```
func shadowsUsedDecl() {
  // This could be anywhere
  let inOuter: String = "str"
  print(inOuter)

  // Refactoring `inner`
  func inner() {
    simpleErr(arg: "str") { inCall, err in
      if inCall != nil {
        let inOuter = inCall!
        print(inOuter)
      }
      // This is the `inOuter` in `shadowsUsedDecl`, not the one in `simpleErr`.
      // Once we hoist `let inOuter = inCall` out though, this would then refer to
      // to the inner one instead. 
      print(inOuter)
    }
  }
}
```

Reply:
Maybe I’m making wrong assumptions about how the async refactoring works here. But don’t we go
- Convert functions decl `inner()` to async alternative
- Walk the body of `inner()`
- Convert the call of `simpleErr` to an async alternative

In which case the entire shadowing is handled by converting the call of `simpleErr` to an async alternative and we don’t need any special handling on the level of function decl conversion at all.